napari/animation/animate.py

```diff
@@ -0,0 +1,336 @@
+
+        for i in range(len(self.states_dict)):
+
+            self.update_viewer_from_state(i)
+            self.viewer.window.qt_viewer.canvas.size = frame_size
+            images.append(self.viewer.screenshot())
+
+        image_stack = np.stack(images, axis=0)
+        return image_stack
+
+    def update_viewer_from_state(self, frame):
+        """Set the viewer to a given interpolated frame
+
+        Parameters
+        -------
+        frame : int
+            frame to visualize
+        """
+
+        # set view type 2D/3D and camera state
+        self.viewer.dims.ndisplay = self.interpolated_states["ndisplay"][frame]
+        self.viewer.window.qt_viewer.view.camera.set_state(
+            self.interpolated_states["camera"][frame]
+        )
+
+        # assign interpolated visibility state
+        for j in range(len(self.viewer.layers)):
+            self.viewer.layers[j].visible = self.interpolated_states["vis"][
+                frame
+            ][j]
+
+        # adjust slider positions
+        for i in range(self.interpolated_states["sliders"].shape[1]):
+            self.viewer.dims.set_point(
+                i, self.interpolated_states["sliders"][frame][i]
+            )
+
+        # update view
+        self.viewer.window.qt_viewer.view.camera.view_changed()
+
+    def make_movie(
+        self,
+        name="movie.mp4",
+        fps=20,
+        resolution=1,
+        compression_quality=5,
+        format=None,
+    ):
+        """Create a movie based on key-frames
+
+        Parameters
+        -------
+        name : str
+            name to use for saving the movie (can also be a path)
+            should be either .mp4 or .gif
+        fps : int
+            frames per second
+        resolution : float
+            factor by which to multiply the current windows size
+        compression_quality: float
+            number from 1 (lowest quality) to 9
+            only applies to mp4
+        format: str
+            The format to use to write the file. By default imageio selects the appropriate for you based on the filename.
+        """
+
+        # creat all states
+        self.create_steps()
+
+        # capture SceneCanvas size of frist frame to set size of next ones
+        self.update_viewer_from_state(0)
+        frame_size = self.viewer.window.qt_viewer.canvas.size
+
+        # create imageio writer and add all frames
+        _, extension = os.path.splitext(name)
+        if extension == '.mp4':
+            writer = imageio.get_writer(
+                name,
+                fps=fps,
+                quality=compression_quality,
+                format=format,
+                output_params=['-timeout', '3'],
+            )
+        else:
+            writer = imageio.get_writer(name, fps=fps, format=format)
+        for frame in range(len(self.interpolated_states["ndisplay"])):
+            self.update_viewer_from_state(frame)
+            self.viewer.window.qt_viewer.canvas.size = (
```

Comment:
I think we want to push this logic into the [viewer.screenshot](https://github.com/napari/napari/blob/master/napari/viewer.py#L65) potentially including adding keyword args to set desired screenshot shape / height / aspect ratio / resolution etc. inside that function that this function can then just also set. I'm not sure if we need all of those to be settable, but maybe having something like `max_shape=None` be a keyword argument so that if `None` is set then it just returns the screenshot as is, and if `max_shape` is an `int` then it will rescale the image so that the largest axis has that length (preserving the aspect ratio).

This function should not ever have to look at the `viewer.window` property or anything inside it, and so if someone passes `max_shape=None` then maybe I'd take a single screenshot, determine the current max shape and then call `self.viewer.screenshot` with `max_shape` set to that value to ensure that the screenshots from that point on always have the right size.

If we go with `max_shape` I'm not sure we then also need a `resolution` keyword argument as I think they are redundant right?

Reply:
Very good idea. This will unclutter the movie function and at the same time offer the resolution option to ```screenshot()``` in general.

Reply:
Small problem here: as mentioned before, when switching from 2D to 3D view, the canvas size is changing. But it's not changing isotropically: additional sliders at the bottom of the viewer are reducing the canvas size *only vertically*. Therefore rescaling is not sufficient. To make a movie all the images need to have the same size, so I actually *need* the ```viewer.window.qt_viewer.canvas.size``` call  in the movie making function to be able to set all frames to the exact same dimensions.

We could of course avoid that if the 2D/3D switch left the canvas size unchanged and rather the viewer itself would change.
Any thoughts on this ?

Reply:
ah hmm, there's not an obvious way to make it work it all cases without forcing the canvas size to be the same, which is something we might want to do. Curious if @tlambert03 has thoughts on that. Regardless as much of this logic as possible should be inside the `screenshot` method.